File: elements-of-programming-interviews/python/advance_by_offsets.py
from typing import List

from test_framework import generic_test


def can_reach_end(A: List[int]) -> bool:
    best = A[0]
    curr = 0
    for dist in A:
        if (curr > best):
            return False
        if (best >= len(A)-1):
            return True
        best = max(curr + A[curr], best)
        curr += 1


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('advance_by_offsets.py',
                                       'advance_by_offsets.tsv',
                                       can_reach_end))

File: elements-of-programming-interviews/2023/cpp/advance_by_offsets.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;
bool CanReachEnd(const vector<int>& arr) {
  int best = arr[0];
  for (int curr = 0; curr < arr.size(); curr++) {
    if (best >= arr.size() - 1) {
      return true;
    }
    if (curr > best) {
      return false;
    }
    best = (curr + arr[curr] > best) ? curr + arr[curr] : best;
  }
  return true;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"max_advance_steps"};
  return GenericTestMain(args, "advance_by_offsets.cc",
                         "advance_by_offsets.tsv", &CanReachEnd,
                         DefaultComparator{}, param_names);
}
